Add configurable groups to bootstrap tokens.

mattmoyer opened the pull request:
**What this PR does / why we need it**:
This change adds support for authenticating bootstrap tokens into a configurable set of extra groups in addition to `system:bootstrappers`. Previously, bootstrap tokens could only ever authenticate to the `system:bootstrappers` group.

Groups are specified as a comma-separated list in the `auth-extra-groups` key of the `bootstrap.kubernetes.io/token` Secret, and must begin with the prefix `system:bootstrapper:` (and match a validation regex that checks against our normal convention). Whether or not any extra groups are configured, `system:bootstrappers` will still be added.

This also adds a `--groups` flag for `kubeadm token create`, which sets the `auth-extra-groups` key on the resulting Secret. The default is to not set the key.

`kubeadm token list` is also updated to include a `EXTRA GROUPS` output column.

**Which issue this PR fixes**: fixes #49306

**Special notes for your reviewer**: 
The use case for this is in https://github.com/kubernetes/kubernetes/issues/49306. Comments on the feature itself are probably better over there. It will be part of how HA/self-hosting kubeadm bootstraps new master nodes (post 1.8).

**Release note**:
```release-note
Add support for configurable groups for bootstrap token authentication.
```

cc @luxas @kubernetes/sig-cluster-lifecycle-api-reviews @kubernetes/sig-auth-api-reviews 

/kind feature

luxas commented on plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go:
```diff
@@ -79,6 +80,7 @@ func tokenErrorf(s *api.Secret, format string, i ...interface{}) {
 //       token-id: ( token id )
 //       # Required key usage.
 //       usage-bootstrap-authentication: true
+//       auth-groups: "group1,group2"
```
`system:bootstrappers:custom-group1.system:bootstrappers:custom-group1` for clarity

luxas commented on plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go:
```diff
@@ -184,3 +192,28 @@ func parseToken(s string) (string, string, error) {
+			continue
+		}
+
+		if !strings.HasPrefix(group, bootstrapapi.BootstrapGroupPrefix) {
```
Should we use a regexp here instead in order to match things better?
This allows things I don't really like:
 - `system:bootstrappers:`
 - `system:bootstrappers:*` (not sure if anything will happen with this, but seems scary)
 - `system:bootstrappers:<script> alert("whoa, some JS inlined here!") </script>`

you get the idea ;)

ericchiang replied:
We don't do any kind of validation like this with other authenticators. It seems more sane to either keep these arbitrary strings or define some some subset of characters the group can use than defend against some possible injections. If you don't trust the value you should sanitize it anyway and shouldn't trust a regexp.

luxas replied:
I propose to just allow `[a-z0-9-:]*`, should cover at least 95% of the cases IMO

mattmoyer replied:
I didn't want to add any restrictions that might be surprising, but I don't think some basic restrictions would be that hard to work with. Maybe something like `\Asystem:bootstrappers:[a-zA-Z0-9:-]{1,255}\z`? 

Is it okay to just put a little `ValidateBootstrapGroupName(name string) error` function into `pkg/bootstrap/api/validate.go` alongside the constants in `types.go`? Exporting the regex as a `const` feels weirder to me than just exporting a function.

luxas replied:
> Exporting the regex as a const feels weirder to me than just exporting a function.

Please export both :)
I think that's what most other APIs do; often a `helpers.go` file is there besides `types.go`

luxas commented on plugin/pkg/auth/authenticator/token/bootstrap/bootstrap_test.go:
```diff
@@ -247,3 +288,75 @@ func TestTokenAuthenticator(t *testing.T) {
+				Data: map[string][]byte{
+					bootstrapapi.BootstrapTokenGroupsKey: []byte(
+						bootstrapapi.BootstrapDefaultGroup + "," +
+							bootstrapapi.BootstrapGroupPrefix + "foo" + "," +
```
nit: weird formatting here?

I think this should be written as it would be in a real BT; as `system:bootstrappers,system:bootstrappers:foo,system:bootstrappers:bar`
If we ever change this prefix it's very reasonable to have to change this unit test.

Cases I want to see covered here:
 - `system:bootstrappers` only
 - `system:bootstrappers:foo` only
 - `system:bootstrappers:foo,system:bootstrappers:bar`
 - `system:bootstrappers,system:bootstrappers:foo,system:bootstrappers:bar` (what you have here)
 - `system:bootstrappers:` (should not be allowed??)
 - `system:bootstrappers:      ` should not be allowed
 - `system:bootstrappers:<script> alert("scary?!") </script>`
 - `system:bootstrappers:foo:` must not end with `:`

etc etc add here whatever you can think of that is useful

luxas commented on plugin/pkg/auth/authenticator/token/bootstrap/bootstrap_test.go:
```diff
@@ -247,3 +288,75 @@ func TestTokenAuthenticator(t *testing.T) {
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Data:       map[string][]byte{},
+			},
+			expectResult: []string{bootstrapapi.BootstrapDefaultGroup},
```
nit: I'm a fan of using hard-coded strings in unit tests sometimes. It seems like some of these could use hard-coding as we want things to be in a certain way. If some constant is changed (unexpectedly), the unit test will fail, and I'm in favor of failing in that case than letting things silently pass

mattmoyer replied:
Ack, that makes sense to me.

luxas commented on plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go:
```diff
@@ -184,3 +192,28 @@ func parseToken(s string) (string, string, error) {
+		}
+
+		if !strings.HasPrefix(group, bootstrapapi.BootstrapGroupPrefix) {
+			return nil, fmt.Errorf("group does not start with %q", bootstrapapi.BootstrapGroupPrefix)
```
nit: `field 'auth-groups' on bootstrap token is invalid, does not start with %q`

mattmoyer replied:
This gets wrapped by the `tokenError` helper above, so the actual warning message is:
`Bootstrap secret %s/%s matching bearer token has invalid value for key auth-groups: group does not start with "system:bootstrappers:".`

luxas commented on pkg/bootstrap/api/types.go:
```diff
@@ -80,6 +88,11 @@ const (
 	BootstrapUserPrefix = "system:bootstrap:"
 
-	// BootstrapGroup is the group bootstrapping bearer tokens authenticate in.
-	BootstrapGroup = "system:bootstrappers"
```
Have you looked up and replaced all occurances of this const? I _think_ there should be some more places...

mattmoyer replied:
This was all I could find.

luxas commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -138,6 +139,9 @@ var (
 	DefaultTokenUsages = []string{"signing", "authentication"}
 
+	// DefaultTokenGroups specifies the default groups a token will get
+	DefaultTokenGroups = []string{bootstrapapi.BootstrapDefaultGroup}
```
See `NodeBootstrapTokenAuthGroup` above. Should use that one; and it should be changed to something like:
 - `system:bootstrappers:kubeadm:node`
 - `system:bootstrappers:kubeadm:default-node`

@ericchiang @liggitt @mikedanese any preference here?

ericchiang replied:
The fewer imports that kubeadm has on k8s.io/kubernetes the easier it's going to be split out. Constants seem like a good candidate for things to just copy/paste and not add another import.

luxas replied:
Isn't it: the less deps the rest of the kubernetes/kubernetes code has on `cmd/kubeadm`; the easier it will be to split out? We already vendor `bootstrapapi` quite naturally, as we're the only user of it currently (AFAIK)

mattmoyer replied:
I ended up not needing the import. I think we should discuss the `NodeBootstrapTokenAuthGroup` change in a second PR because that potentially breaks some upgraded clusters if I'm understanding correctly.

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -208,8 +212,18 @@ func RunCreateToken(out io.Writer, client clientset.Interface, token string, tok
 	}
 
+	// validate groups
+	for _, group := range groups {
```
Please add a TODO that as soon as https://github.com/kubernetes/client-go/issues/114 is fixed; we should get rid of this.
This should also be synced with what we do below and splitted out to its own func.

luxas reviewed (commented):
Thanks @mattmoyer :tada:!

I left a couple of review comments below, please fix them.
In particular, I think we should have a regex that only allows `[:a-z]` (possibly `[0-9]`?) after `system:bootstrappers`. We should also think about a maximum length of this group... 255 chars max just as it is a nice number :)?

@liggitt do we have a max length on groups in RBAC already or pre-made validation funcs for a group?

@mattmoyer also I want the `cmd/kubeadm`, `pkg` and `plugin` changes in separate commits for cleanliness.

ericchiang commented on plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go:
```diff
@@ -184,3 +192,28 @@ func parseToken(s string) (string, string, error) {
+	for _, group := range groups {
+		if group == bootstrapapi.BootstrapDefaultGroup {
+			// allow the default group even though it doesn't strictly match the prefix
+			continue
```
Would have expected this to behave like `system:authenticated` and always be added. E.g. if I specify `"foo,bar"` I get `["system:bootstrappers","system:bootstrappers:foo","system:bootstrappers:bar"]`

mattmoyer replied:
Ah, I can do that. Maybe rename `auth-groups` to `extra-auth-groups` to make it clear that they're an addition?

luxas replied:
@mattmoyer SGTM :+1:

mattmoyer replied:
I did this sort of halfway. I made it so `system:bootstrappers` is always added for bootstrap tokens, but I still made it so the serialized values in the `auth-extra-groups` key includes the full prefix, so if you set the key to `"system:bootstrappers:foo,system:bootstrappers:bar"` you get three groups:
 - `system:bootstrappers`
 - `system:bootstrappers:foo`
 - `system:bootstrappers:bar`

@ericchiang does that work for you? I prefer for the serialized value to be very explicit so there's just validation but not transformation happening in the authenticator.

mattmoyer committed:
Add extra group constants and validation to `pkg/bootstrap/api`.

This adds constants and validation for a new `auth-extra-groups` key on `bootstrap.kubernetes.io/token` secrets. This key allows a bootstrap token to authenticate to extra groups in addition to the `system:bootstrappers` group.

Extra groups are always applied in addition to the `system:bootstrappers` group, must begin with a `system:bootstrappers:` prefix, are limited in length, and are limited to a restricted set of characters (alphanumeric, colons, and dashes without a trailing colon/dash).

mattmoyer committed:
Implement `auth-extra-groups` in bootstrap token authenticator.

This implements support for the new `auth-extra-groups` key in `bootstrap.kubernetes.io/token` secrets by adding extra groups to the user info returned for valid bootstrap tokens.

mattmoyer committed:
kubeadm: add `--groups` flag for `kubeadm token create`.

This adds support for creating a bootstrap token that authenticates with extra `system:bootstrappers:*` groups in addition to `system:bootstrappers`.

mattmoyer commented:
Thanks for the reviews, everyone. I believe I've addressed the initial round of comments, so please take another look.

ericchiang commented on plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go:
```diff
@@ -79,6 +80,7 @@ func tokenErrorf(s *api.Secret, format string, i ...interface{}) {
 //       token-id: ( token id )
 //       # Required key usage.
 //       usage-bootstrap-authentication: true
+//       auth-extra-groups: "system:bootstrappers:custom-group1.system:bootstrappers:custom-group1"
```
Is the "." separator supposed to be ","?

mattmoyer replied:
Yes. Thanks!

mattmoyer commented:
/retest

luxas commented:
@liggitt @ericchiang PTAL

ericchiang commented:
/lgtm

mattmoyer committed:
kubeadm: add extra group info to `token list`.

This adds an `EXTRA GROUPS` column to the output of `kubeadm token list`. This displays any extra `system:bootstrappers:*` groups that are specified in the token's `auth-extra-groups` key.

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -281,7 +281,7 @@ func (i *Init) Run(out io.Writer) error {
 	// Create the default node bootstrap token
 	tokenDescription := "The default bootstrap token generated by 'kubeadm init'."
-	if err := nodebootstraptokenphase.UpdateOrCreateToken(client, i.cfg.Token, false, i.cfg.TokenTTL, kubeadmconstants.DefaultTokenUsages, tokenDescription); err != nil {
+	if err := nodebootstraptokenphase.UpdateOrCreateToken(client, i.cfg.Token, false, i.cfg.TokenTTL, kubeadmconstants.DefaultTokenUsages, []string{}, tokenDescription); err != nil {
```
The default kubeadm token should have a prefix of `system:bootstrappers:kubeadm:default-node` or something.
Please add that to constants.go
We should also migrate the RBAC-creation code to create a ClusterRoleBinding depending on version

mattmoyer replied:
Are you okay doing that in a followup PR? I wanted to make sure we're careful about upgrading the cluster role binding since it's possible that an upgraded cluster would have nodes using tokens created without the new default group.

luxas replied:
I'm ok with that.

liggitt replied:
> The default kubeadm token should have a prefix of system:bootstrappers:kubeadm:default-node or something.

That seems like you're using a group to represent an individual token... each token already has its own unique username... why not use that?

mattmoyer replied:
I think there's a slight difference where you might have a single logical node pool (maybe `system:bootstrappers:high-integrity-nodes`). That group name might be used in a custom CSR autoapprover to assign a particular class of node identity to those nodes.

You might rotate through multiple tokens for different nodes within that pool, e.g., a token you used to bootstrap an initial set of nodes, and then another token you use to bootstrap additional nodes later on.

luxas replied:
So this is for the default token _group_, sorry for not making that clear.
ClusterRoleBindings will be bound to this group; and the default token will be added to this group by default.

liggitt commented on pkg/bootstrap/api/types.go:
```diff
@@ -80,6 +86,12 @@ const (
+	// BootstrapGroupPattern is the valid regex pattern that all groups
+	// assigned to a bootstrap token by BootstrapTokenExtraGroupsKey must match.
+	// See also ValidateBootstrapGroupName().
+	BootstrapGroupPattern = "system:bootstrappers:[a-zA-Z0-9:-]{0,255}[a-zA-Z0-9]"
```
is there a reason to use mixed case? lowercase is simpler and avoids typos

mattmoyer replied:
I can make this more restrictive if we want. These groups will probably primarily come from things we generate inside kubeadm, where we follow the `kebab-case` convention used by all the other group names.

 @luxas does that work for you?

Maybe:
`"system:bootstrappers:[a-z0-9:-]{0,255}[a-z0-9]"` (no uppercase)

Or even:
`"system:bootstrappers:[a-z:-]{0,255}[a-z]"` (no numbers either)

liggitt replied:
> kebab-case convention used by all the other group names

examples of this? I thought all our group names were lowercase alpha currently

luxas replied:
> "system:bootstrappers:[a-z0-9:-]{0,255}[a-z0-9]" (no uppercase)

This is what I (think I) proposed above, SGTM

liggitt commented on plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go:
```diff
@@ -184,3 +192,24 @@ func parseToken(s string) (string, string, error) {
+// getGroups loads and validates the bootstrapapi.BootstrapTokenExtraGroupsKey
+// key from the bootstrap token secret, returning a list of group names or an
+// error if any of the group names are invalid.
+func getGroups(secret *api.Secret) ([]string, error) {
```
accumulating into a sets.NewString() would dedupe automatically for you

```go
groups := sets.NewString(bootstrapapi.BootstrapDefaultGroup)

if additionalGroups := getSecretString(secret, bootstrapapi.BootstrapTokenExtraGroupsKey); len(additionalGroups) > 0 {
  ...
} 

return groups.List(), nil
```

mattmoyer replied:
+1, de-duping seems like a good idea.

liggitt commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -213,8 +217,14 @@ func RunCreateToken(out io.Writer, client clientset.Interface, token string, tok
 		}
 	}
 
+	for _, group := range extraGroups {
```
should probably only allow groups if usages has BootstrapTokenUsageAuthentication, right?

mattmoyer replied:
+1, it's just a client side check but it's good validation so we can fail fast with a nice error message.

mattmoyer commented:
Updated to address @liggitt's comments. PTAL.

I also added a fourth commit that adds the extra group info to the output of `kubeadm token list`.

mattmoyer commented:
/retest

mattmoyer commented:
/retest

liggitt commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -246,7 +265,7 @@ func RunListTokens(out io.Writer, errW io.Writer, client clientset.Interface) er
 
 	w := tabwriter.NewWriter(out, 10, 4, 3, ' ', 0)
-	fmt.Fprintln(w, "TOKEN\tTTL\tEXPIRES\tUSAGES\tDESCRIPTION")
+	fmt.Fprintln(w, "TOKEN\tTTL\tEXPIRES\tUSAGES\tDESCRIPTION\tEXTRA GROUPS")
```
nit: column headers shouldn't have whitespace in the titles

liggitt commented:
/lgtm

luxas commented:
/approve

jbeda commented on pkg/bootstrap/api/types.go:
```diff
@@ -80,6 +86,12 @@ const (
+	// BootstrapGroupPattern is the valid regex pattern that all groups
+	// assigned to a bootstrap token by BootstrapTokenExtraGroupsKey must match.
+	// See also ValidateBootstrapGroupName().
+	BootstrapGroupPattern = "system:bootstrappers:[a-z0-9:-]{0,255}[a-z0-9]"
```
Nit: usually for stuff like this we use a DNS-ish type of regexp.  That means that any segment shouldn't start with a number or -. It is a bit strange to allow `system:bootstrappers:-foo` here, for example.

But I don't see any real problem to be honest so it is just a nit.

jbeda commented:
/approve

Also don't forget to do the doc updates.  This'll hit the bootstrap authn page and the kubeadm reference page.  There might be a few more pages to reference this in.

luxas commented:
/retest

mattmoyer commented:
/retest

mattmoyer commented:
Failing test looks like https://github.com/kubernetes/kubernetes/issues/51429.
